src/providers/twitch/PubSubClient.cpp

```diff
@@ -1,224 +1,192 @@
-    DebugCount::increase("PubSub topic pending listens",
-                         static_cast<int64_t>(numRequestedListens));
 
-    for (const auto &topic : msg.topics)
+    auto message = *optMessage;
+
+    switch (message.type)
     {
-        this->listeners_.emplace_back(Listener{
-            TopicData{
-                topic,
-                false,
-                false,
-            },
-            false,
-        });
-    }
+        case PubSubMessage::Type::Pong: {
+            this->lastHeartbeat_.store(std::chrono::steady_clock::now());
+        }
+        break;
 
-    qCDebug(chatterinoPubSub)
-        << "Subscribing to" << numRequestedListens << "topics";
+        case PubSubMessage::Type::Response: {
+            this->handleResponse(message);
+        }
+        break;
 
-    this->send(msg.toJson());
+        case PubSubMessage::Type::Message: {
+            auto oMessageMessage = message.toInner<PubSubMessageMessage>();
+            if (!oMessageMessage)
+            {
+                qCDebug(chatterinoPubSub) << "Malformed MESSAGE:" << msg;
+                return;
+            }
+
+            this->handleMessageResponse(*oMessageMessage);
+        }
+        break;
 
-    return true;
+        case PubSubMessage::Type::INVALID:
+        default: {
+            qCDebug(chatterinoPubSub)
+                << "Unknown message type:" << message.typeString;
+        }
+        break;
+    }
 }
 
-PubSubClient::UnlistenPrefixResponse PubSubClient::unlistenPrefix(
-    const QString &prefix)
+void PubSubClient::checkHeartbeat()
 {
-    std::vector<QString> topics;
-
-    for (auto it = this->listeners_.begin(); it != this->listeners_.end();)
+    if (!this->isOpen_)
     {
-        const auto &listener = *it;
-        if (listener.topic.startsWith(prefix))
-        {
-            topics.push_back(listener.topic);
-            it = this->listeners_.erase(it);
-        }
-        else
-        {
-            ++it;
-        }
+        return;
     }
 
-    if (topics.empty())
+    auto dur = std::chrono::steady_clock::now() - this->lastHeartbeat_.load();
+    if (dur > this->heartbeatInterval_ * 1.5)
     {
-        return {{}, ""};
+        qCDebug(chatterinoPubSub) << "Heartbeat timed out";
+        this->close();
     }
 
-    auto numRequestedUnlistens = topics.size();
-
-    this->numListens_ -= numRequestedUnlistens;
-    DebugCount::increase("PubSub topic pending unlistens",
-                         static_cast<int64_t>(numRequestedUnlistens));
-
-    PubSubUnlistenMessage message(topics);
-
-    this->send(message.toJson());
-
-    return {message.topics, message.nonce};
+    this->sendText(R"({"type":"PING"})"_ba);
 }
 
-void PubSubClient::handleListenResponse(const PubSubMessage &message)
+QByteArray PubSubClient::encodeSubscription(const Subscription &subscription)
 {
+    PubSubListenMessage listen({subscription.topic});
+    this->nonces_[listen.nonce] = NonceInfo{.isListen = true};
```

Comment:
just formatting nit (will have to be autoformatted after)
```suggestion
    this->nonces_[listen.nonce] = NonceInfo{.isListen = true,};
```